### PR TITLE
[TorchAX][DeepSeek JAX] Enable MLA for TorchAX + Consolidate MLA Kernel Call

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

In this PR, we do the following:

* Move the MLA kernel wrapper from within DeepSeek to a shared interface
* Enable usage of MLA in TorchAX

# Tests

## TorchAX accuracy / perf on MMLU
CMDs:
```
MOE_REQUANTIZE_WEIGHT_DTYPE=fp4 MOE_REQUANTIZE_BLOCK_SIZE=512 NEW_MODEL_DESIGN=1 MODEL_IMPL_TYPE=vllm nohup vllm serve deepseek-ai/DeepSeek-R1 --max-model-len=4096 --max-num-batched-tokens=4096 --max-num-seqs=512 --kv-cache-dtype=fp8 --no-enable-prefix-caching --gpu-memory-utilization=0.9 --tensor-parallel-size=8  --async-scheduling --enable-expert-parallel --additional_config='{"sharding": {"sharding_strategy": {"enable_dp_attention": true}}}' &

python scripts/vllm/benchmarking/benchmark_serving.py --backend vllm --model=deepseek-ai/DeepSeek-R1 --dataset-name mmlu --dataset-path /mnt/disks/jacobplatin/mmlu/data/test/ --num-prompts 1000 --run_eval --temperature 0
```

Results:
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  35.66     
Total input tokens:                      162129    
Total generated tokens:                  2000      
Request throughput (req/s):              28.04     
Output token throughput (tok/s):         56.08     
Total token throughput (tok/s):          4602.07   
---------------Time to First Token----------------
Mean TTFT (ms):                          25133.46  
Median TTFT (ms):                        23179.25  
P99 TTFT (ms):                           35530.46  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          4825.22   
Median TPOT (ms):                        6192.97   
P99 TPOT (ms):                           6219.43   
---------------Inter-token Latency----------------
Mean ITL (ms):                           4825.22   
Median ITL (ms):                         6192.97   
P99 ITL (ms):                            6219.43   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          29958.68  
Median E2EL (ms):                        29397.45  
P99 E2EL (ms):                           35587.41  
==================================================
Evaluating MMLU...
[nltk_data] Downloading package punkt to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!

Results

{'accuracy': 0.854, 'gen_num': 1000}
```

## JAX accuracy / perf on MMLU
CMDs:
```
USE_2D_TP=1 USE_UNFUSED_MEGABLOCKS=1  NEW_MODEL_DESIGN=1 TPU_BACKEND_TYPE=jax  vllm serve --model=jrplatin/DeepSeek-R1-1D-Subchannel-256-FP8 --max-model-len=4096 --tensor-parallel-size 2 --max-num-seqs 16 --max-num-batched-tokens 256 --no-enable-prefix-caching --disable-log-requests --gpu-memory-utilization 0.95 --additional_config='{"sharding": {"sharding_strategy": {"enable_dp_attention": true, "expert_parallelism": 4, "tensor_parallelism": 2}}, "sparse_matmul": "True"}' --hf_overrides '{"architectures": ["DeepseekV3ForCausalLM"]}' --kv-cache-dtype=fp8 --async-scheduling 

 python scripts/vllm/benchmarking/benchmark_serving.py --backend vllm --model=jrplatin/DeepSeek-R1-1D-Subchannel-256-FP8 --dataset-name mmlu --dataset-path /mnt/disks/jacobplatin/mmlu/data/test/ --num-prompts 1000 --run_eval --temperature 0
```

Results:

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  62.33     
Total input tokens:                      162129    
Total generated tokens:                  2000      
Request throughput (req/s):              16.04     
Output token throughput (tok/s):         32.09     
Total token throughput (tok/s):          2633.23   
---------------Time to First Token----------------
Mean TTFT (ms):                          36349.40  
Median TTFT (ms):                        38200.16  
P99 TTFT (ms):                           61446.35  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          744.84    
Median TPOT (ms):                        751.25    
P99 TPOT (ms):                           753.50    
---------------Inter-token Latency----------------
Mean ITL (ms):                           744.84    
Median ITL (ms):                         751.25    
P99 ITL (ms):                            753.50    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          37094.25  
Median E2EL (ms):                        38953.67  
P99 E2EL (ms):                           62190.23  
==================================================
Evaluating MMLU...
[nltk_data] Downloading package punkt to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/jacobplatin_google_com/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!

Results

{'accuracy': 0.845, 'gen_num': 1000}
```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Review by lk-chen — APPROVED** [review]

**Review by kyuyeunk — CHANGES_REQUESTED** [review]
chatted offline & requested to also use attention registering mechanism provided by vllm for better compatibility with existing attention / kv cache management code.

File: tests/layers/vllm/test_mla_attention.py
Comment (jrplatin) [inline tests/layers/vllm/test_mla_attention.py:1]: Need to fix a failing test here (test itself is failing, not the underlying logic) but core logic should be good

File: tpu_inference/platforms/tpu_platform.py [hunk 1]
```diff
@@ -66,7 +66,10 @@ def get_attn_backend_cls(cls, selected_backend: "AttentionBackendEnum",
 
         # Invoke @register_backend in the module.
         import tpu_inference.layers.vllm.attention  # noqa: F401
-        if selected_backend != AttentionBackendEnum.FLASH_ATTN:
+        use_mla = attn_selector_config.use_mla
+        if use_mla:
+            selected_backend = AttentionBackendEnum.FLASHMLA
```
Comment (kyuyeunk) [inline tpu_inference/platforms/tpu_platform.py]: nit: can you change this to `FLASH_ATTN_MLA` for consistency with below's `FLASH_ATTN`?: https://github.com/vllm-project/vllm/blob/da543d1abe2468a1b79f230e91e8bbdc2bf6ee71/vllm/v1/attention/backends/registry.py#L75C5-L75C19

File: tpu_inference/runner/kv_cache_manager.py [hunk 2]
```diff
@@ -150,8 +151,12 @@ def get_kv_cache_spec(self):
                                 block_size, num_kv_heads, head_size)
         else:
             # Else propagate attention modules from compilation config.
-            layers = get_layers_from_vllm_config(self.runner.vllm_config,
-                                                 Attention)
+            attn = get_layers_from_vllm_config(self.runner.vllm_config,
```
Comment (kyuyeunk) [inline tpu_inference/runner/kv_cache_manager.py]: nit: create a list `[Attention, MLAAttention]` and use a for loop to iterate them and add to the dictionary. This will make it easy to extend when new we need to support new attention type.

File: tpu_inference/layers/vllm/mla_attention.py
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/mla_attention.py:1]: can you create a separate folder called `backends` can put both `flash_attn_mla.py` (renamed from `mla_attention.py`)and `flasn_attn.py` (renamed from existing `attention.py`)?

File: tpu_inference/layers/vllm/mla_attention.py [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+    @staticmethod
+    def get_name() -> str:
+        return "FLASHMLA"
+
+    @staticmethod
+    def get_impl_cls() -> type["PallasMLAttentionBackend"]:
+        return PallasMLAttentionBackendImpl
+
+
+class TPUMLAAttention(MLAAttention):
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/mla_attention.py]: note that vLLM makes distinction between `backends` (stateless functional code that has logic for attention forward pass) vs. `attention layers` (stateful attention layer that has weights, stores parameters, and such). so while above `PallasMLAAttentionBackend` should go into `backends` folder, this class should remain in `mla_attention.py` (though it might be better to move this into `layers/vllm/attention/mla_attention.py`?)

File: tpu_inference/layers/vllm/mla_attention.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+                                                  kv_cache, attn_metadata,
+                                                  mesh, self, **kwargs)
+
+        # Update KV cache
+        vllm_model_wrapper_context.kv_caches[kv_cache_index] = new_kv_cache
+
+        return torch_view(outputs)
+
+
+class TPUMultiHeadLatentAttentionWrapper(MultiHeadLatentAttentionWrapper):
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/mla_attention.py]: nit: change this to `VllmMultiHead...` to be consistent with other class names.

File: tpu_inference/layers/vllm/mla_attention.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+            prefix=f"{prefix}.attn",
+            kv_b_proj=self.kv_b_proj,
+            use_sparse=self.is_sparse,
+            indexer=self.indexer,
+        )
+
+        self.prefix = prefix
+
+
+def _jax_mla_func(
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/mla_attention.py]: what is the purpose of this code when we can just immediately call `mla_attention` function?

File: tpu_inference/layers/vllm/mla_attention.py [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+        cache_config: CacheConfig | None = None,
+        quant_config: QuantizationConfig | None = None,
+        prefix: str = "",
+        use_sparse: bool = False,
+        indexer: object | None = None,
+        **extra_impl_args,
+    ):
+        torch.nn.Module.__init__(self)
+
+        self.num_heads = num_heads
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/mla_attention.py]: how much of this initialization code needs to be implemented here vs. how much we can reuse from vllm's `MLAAttention` class? https://github.com/vllm-project/vllm/blob/da543d1abe2468a1b79f230e91e8bbdc2bf6ee71/vllm/model_executor/layers/attention/mla_attention.py#L299-L419

File: tpu_inference/layers/vllm/mla_attention.py (hunk 5 above)
Comment (jrplatin) [reply tpu_inference/layers/vllm/mla_attention.py]: Fair point -- was trying to be consistent with how [non-MLA attention](https://github.com/vllm-project/tpu-inference/blob/b4a6a023d0e8e7eecb08a1628e75cf63fc69a167/tpu_inference/layers/vllm/attention.py#L234) is constructed, but makes sense just to simplify the code

File: tpu_inference/layers/vllm/mla_attention.py
Comment (jrplatin) [reply tpu_inference/layers/vllm/mla_attention.py:1]: Yep!

File: tpu_inference/layers/vllm/mla_attention.py (hunk 3 above)
Comment (jrplatin) [reply tpu_inference/layers/vllm/mla_attention.py]: Got it -- so just to confirm: let's move `PallasMLAttentionBackendImpl` to `layers/vllm/backends/flash_attn_mla.py` but keep the rest of the code in `layers/vllm/mla_attention.py`

For the current `layers/vllm/attention.py` code, we'll move all of it (since it's all `Impl`/backend-based code) to `layers/vllm/backends/flash_attn.py`. correct?

File: tpu_inference/layers/vllm/mla_attention.py (hunk 6 above)
Comment (jrplatin) [reply tpu_inference/layers/vllm/mla_attention.py]: Was able to consolidate, good call out, thanks!

File: tpu_inference/layers/vllm/mla_attention.py (hunk 5 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/mla_attention.py]: yeah, honestly, i wanted to remove that code in non-mla attention as well. so if you want to do that as well in this pr, that would great!

File: tpu_inference/layers/vllm/mla_attention.py (hunk 3 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/mla_attention.py]: `PallasMLAttentionBackendImpl` and `PallasMLAttentionBackend` to be accurate. but yup. that's correct.

File: tpu_inference/layers/vllm/mla_attention.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                                            MultiHeadLatentAttentionWrapper)
+from vllm.model_executor.layers.quantization import QuantizationConfig
+from vllm.v1.attention.backend import AttentionType
+
+from tpu_inference import utils
+from tpu_inference.models.vllm.vllm_model_wrapper_context import \
+    get_vllm_model_wrapper_context
+
+
+class TPUMLAAttention(MLAAttention):
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/mla_attention.py]: can you rename this to `VllmMLAAttention` as well?

File: tpu_inference/platforms/tpu_platform.py [hunk 8]
```diff
@@ -65,8 +65,12 @@ def get_attn_backend_cls(cls, selected_backend: "AttentionBackendEnum",
         from vllm.v1.attention.backends.registry import AttentionBackendEnum
 
         # Invoke @register_backend in the module.
-        import tpu_inference.layers.vllm.attention  # noqa: F401
-        if selected_backend != AttentionBackendEnum.FLASH_ATTN:
+        import tpu_inference.layers.vllm.backends.flash_attn  # noqa: F401
```
Comment (kyuyeunk) [inline tpu_inference/platforms/tpu_platform.py]: nit: create `layers/vllm/backends/__init__.py` that imports these functions so that `tpu_platform.py` only needs to import that file.
